test/unit/strip_level/strip_level_test_suite.bzl

```diff
@@ -0,0 +1,100 @@
+"""Starlark tests for `rust_toolchain.strip_level`"""
+
+load("@bazel_skylib//lib:unittest.bzl", "analysistest")
+load("@bazel_skylib//rules:write_file.bzl", "write_file")
+load("//rust:defs.bzl", "rust_binary")
+load(
+    "//test/unit:common.bzl",
+    "assert_action_mnemonic",
+    "assert_argv_contains",
+)
+
+def _strip_level_test_impl(ctx, expected_level):
+    env = analysistest.begin(ctx)
+    target = analysistest.target_under_test(env)
+
+    action = target.actions[0]
+    assert_action_mnemonic(env, action, "Rustc")
+
+    assert_argv_contains(env, action, "--codegen=strip={}".format(expected_level))
+    return analysistest.end(env)
+
+def _strip_level_for_dbg_test_impl(ctx):
+    return _strip_level_test_impl(ctx, "none")
+
+_strip_level_for_dbg_test = analysistest.make(
+    _strip_level_for_dbg_test_impl,
+    config_settings = {
+        "//command_line_option:compilation_mode": "dbg",
+    },
+)
+
+def _strip_level_for_fastbuild_test_impl(ctx):
+    return _strip_level_test_impl(ctx, "none")
+
+_strip_level_for_fastbuild_test = analysistest.make(
+    _strip_level_for_fastbuild_test_impl,
+    config_settings = {
+        "//command_line_option:compilation_mode": "fastbuild",
+    },
+)
+
+def _strip_level_for_opt_test_impl(ctx):
+    return _strip_level_test_impl(ctx, "debuginfo")
+
+_strip_level_for_opt_test = analysistest.make(
+    _strip_level_for_opt_test_impl,
+    config_settings = {
+        "//command_line_option:compilation_mode": "opt",
+    },
+)
+
+def strip_level_test_suite(name):
+    """Entry-point macro called from the BUILD file.
+
+    Args:
+        name (str): The name of the test suite.
+    """
+    write_file(
+        name = "bin_main",
+        out = "main.rs",
+        content = [
+            "fn main() {}",
+            "",
+        ],
+    )
+
+    rust_binary(
+        name = "bin",
+        srcs = [":main.rs"],
+        edition = "2021",
+        target_compatible_with = [
```

Comment:
What is the failure on windows? I would expect this to "just work" given that this PR is adding plumbing for an existing `rustc` flag.

Reply:
`(13:36:05) ERROR: C:/b/wmjdfsig/external/rules_rust/util/process_wrapper/BUILD.bazel:31:36: output 'external/rules_rust/util/process_wrapper/process_wrapper.pdb' was not created`

Reply:
```
(13:36:16) WARNING: Remote Cache: Expected output util/process_wrapper/process_wrapper.pdb was not created locally.
java.io.IOException: Expected output util/process_wrapper/process_wrapper.pdb was not created locally.
	at com.google.devtools.build.lib.remote.RemoteExecutionService.lambda$buildUploadManifestAsync$6(RemoteExecutionService.java:1339)
	at io.reactivex.rxjava3.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:43)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
	at io.reactivex.rxjava3.core.Single.blockingGet(Single.java:3644)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.buildUploadManifest(RemoteExecutionService.java:1365)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.uploadOutputs(RemoteExecutionService.java:1420)
	at com.google.devtools.build.lib.remote.RemoteSpawnCache$1.store(RemoteSpawnCache.java:188)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:164)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:119)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.analysis.actions.SpawnAction.execute(SpawnAction.java:261)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1144)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1061)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:558)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:859)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:333)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:171)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
(13:36:16) ERROR: C:/b/bk-windows-x2t8/bazel/rules-rust-rustlang/util/process_wrapper/BUILD.bazel:31:36: output 'util/process_wrapper/process_wrapper.pdb' was not created
(13:36:16) ERROR: C:/b/bk-windows-x2t8/bazel/rules-rust-rustlang/util/process_wrapper/BUILD.bazel:31:36: Compiling Rust (without process_wrapper) bin process_wrapper (6 files) [for tool] failed: not all outputs were created or valid
(13:36:16) ERROR: C:/b/bk-windows-x2t8/bazel/rules-rust-rustlang/util/process_wrapper/BUILD.bazel:31:36: output 'util/process_wrapper/process_wrapper.pdb' was not created
(13:36:16) ERROR: C:/b/bk-windows-x2t8/bazel/rules-rust-rustlang/util/process_wrapper/BUILD.bazel:31:36: Compiling Rust (without process_wrapper) bin process_wrapper (6 files) [for tool] failed: not all outputs were created or valid
```

Reply:
https://buildkite.com/bazel/rules-rust-rustlang/builds/11059#018e9f04-6125-4737-b6a6-9cb63922c0af

here is a build with the windows failures.

Reply:
I think this means there's an additional change required to detect when stripping would be enabled and prevent the capture of the `.pdb` file on Windows which seems to not get produced when strip-debug-info is enabled.

Reply:
It's not if it's an opt build, right? It's if stripping debug info is enabled? Can you see if this fails even for `dbg` builds if you set
```python
strip_level = {
    "opt": "debug_info",
    "fastbuild": "debug_info",
    "dbg": "debug_info",
}
```

Also, now that I think about it, are there more levels? Maybe you could link to some rustc docs?

Reply:
https://github.com/bazelbuild/rules_rust/pull/2513/commits/59bf5a640aa6d0d1965d648cdd32d032f42bb79f

Reply:
https://buildkite.com/bazel/rules-rust-rustlang/builds/11060#018e9f1c-bba4-4b8a-a5ca-d6a833a62074

Same errors which makes sense. its trying to declare a file thats not being produced.

Reply:
Yeah, I think at the very least the defaults should be expected to work on all platforms. So we should fix the windows issue in this change.

Reply:
See the Rust 1.77.1 release notes where this change is disabled on Windows in Cargo.

Reply:
@riking is this going to break peoples workflow? We could just not apply the strip flags to windows